### ISSUE TITLE
Clarify variable scoping

### ISSUE THREAD

**Issue by jakearchibald** [whatwg/infra#204]
Currently the spec says "block scope", but can we be more specific. For example: variable declarations are scoped to the closest ordered list or algorithm.

This would make this example from @tobie invalid:

```
1.  If condition, then let |result| be `"foo"`.
1.  Otherwise, let |result| be `"bar"`.
```

But I think it's better written as:

```
1. Let |result| be `"foo"` if condition, otherwise `"bar"`.
```

Step 3 of https://fetch.spec.whatwg.org/#http-network-fetch would also need tweaking.

Also, I think:

> Variables must not be declared more than once per algorithm.

Would be more clearly written as:

> Variable names must be unique within an algorithm.

I didn't realise the original statement was to prevent shadowing.

**Comment by tabatkins** [comment]
Specifically, I think the things in HTML that should be considered "blocks" are:

* the actual algorithm container, obvs
* `<ol>`
* `<dd>`
* maybe `<ul>`? Or perhaps the `<li>` within it? Or maybe neither should establish variable scopes.

Is there anything else?

**Comment by annevk** [comment]
I don't think `<ul>` counts.

I wonder how to handle step 3 in https://url.spec.whatwg.org/#concept-ipv4-serializer. Ideally step 4 doesn't have access to _i_, right?

**Comment by tobie** [comment]
So we'd need to define:

- what a block is,
- how to pass blocks an initial set of key/value pairs (variable name and value) which are declared in it,
- that variables are accessible _by reference_ from nested steps or following steps (i.e. that variable declarations are not hoisted).

A new block is created:

1. implicitly every time a _desugared_ step is nested,
1. explicitly, for certain constructs (e.g. loops).

Conditional statements which occur in the same step are—I think—the only ones which need to be 
desugared for now:
 
```
1.  If condition, then let |result| be `"foo"`.
```

Which is in fact sugar for:

```
1.  If condition, then:
    1.  let |result| be `"foo"`.
```

Blocks in loops are  the only ones which need to take initial key/value pairs (usually each item of the list that's being looped over) and be explicitly constructed. For example, for "For each":

```html
<p>To <dfn export for=list,set lt="iterate|for each">iterate</dfn> over a <a>list</a>, performing a
set of steps on each <a for=list>item</a> in order, use phrasing of the form
"<a for=list>For each</a> |item| of <var ignore>list</var>", and then operate on |item| in the
subsequent prose.
```

We should add something like:

```html
<p><a for=list>For each</a> creates a new <a>block</a> for each <a>item</a> in <a>list</a>,
with its <a>initial variable</a> "item" set to the corresponding <a>item</a>.
```
So we'd have to add a definition of _block_, _initial variable_, and _desugared_ step.

**Comment by tabatkins** [comment]
> Ideally step 4 doesn't have access to i, right?

Ah, yes, Infra looping constructs should create blocks on their own.  (We don't want Python-style visibility of the final iteration's variable bindings to later steps.)

> Conditional statements which occur in the same step are—I think—the only ones which need to be desugared for now:

I don't think this is correct; I think you'll usually want |foo| to be visible to step 2 of that algorithm, not be scoped to just that item.

**Comment by tobie** [comment]
>> Conditional statements which occur in the same step are—I think—the only ones which need to be desugared for now:
> I don't think this is correct; I think you'll usually want |foo| to be visible to step 2 of that algorithm, not be scoped to just that item.

I'm torn. On one side it seems like we really want the equivalent of a ternary operator here (as @jakearchibald mentioned in the OP). On the other it feels very weird (and error prone when refactoring) that the two following statements wouldn't be the same.

```
1.  If condition, then let |result| be `"foo"`.
```

```
1.  If condition, then:
    1.  let |result| be `"foo"
```

**Comment by tabatkins** [comment]
I don't think that's particularly error-prone; presumably there will be more steps in the inner list.  If you want that variables exposed in the outer list, you'll define it in a previous step, then use `set` in the inner list to update the value.  And with the scoping rules in place, it'll be obvious when you mess up. (If you refer to the definition outside the inner list, it'll complain; if you use `let` in the inner list *and* the outer list, it'll complain.)

**Comment by tobie** [comment]
@tabatkins, to clarify, your suggestion is that we don't need to introduce desugaring at all? If so, what do you think should happen at step 3 and what should be the value of |result| in that case?

```
1.  Let |condition| be false.
1.  If |condition|, then let |foo| be `"foo"`.
1.  Let |result| be |foo|.

**Comment by tabatkins** [comment]
Yes, I don't think we need to do anything special for "if" statements.

That code example should be fine - the step 3 |foo| is referencing the |foo| defined in step 2.  The fact that |foo| is undefined if |condition| is false is a spec bug, that's all.

(I can see the argument for making "if" a scoping block, but I can't auto-detect that since conditionals aren't Infra constructs.)

**Comment by tobie** [comment]
> (I can see the argument for making "if" a scoping block, but I can't auto-detect that since conditionals aren't Infra constructs.)

What are your plans for handling |item| in `For each |item| in |list|`, then?

**Comment by tabatkins** [comment]
Only trigger the proper scoping behavior when people do `[=For each=] |item| in |list|` (that is, if the first node in the element is a link to Infra's for-each).

**Comment by jakearchibald** [comment]
It's probably easiest to say that, with `[=For each=] |item|`, *item* is scoped to the `<li>`.

That means both of these work:

```
1. [=For each=] |item| of |list|, do _something_ with |list|.
2.[=For each=] |item| of |list|:
  1. Do _something_ with |list|. 
```

**Comment by tabatkins** [comment]
Yes, agree.

**Comment by jakearchibald** [comment]
Hmm, my idea doesn't work in this case:

```
1. [=For each=] |item| of |list|:
  1. Run the following steps [=in parallel=]:
    1. Do _something_ with |item|. 
```

Doesn't seem like we can define for-each scoping using elements.

**Comment by jakearchibald** [comment]
I guess behaviour could switch depending on whether steps are provided or not.

**Comment by annevk** [comment]
> Variable names must be unique within an algorithm.

Is this actually more clear? At least to me this reads as if you cannot have the same name occur more than once, which is contradictory with the purpose of variables, but that's how it reads.

**Comment by jakearchibald** [comment]
Hmm, I see what you mean.

Is this intent here to prevent shadowing, or to prevent any kind of reuse?

When writing JS, some folks don't like shadowing, but I don't think folks are against reuse:

```js
const links = [...document.querySelectorAll('a[href]')];
const crossOriginLinks = links.map(a => new URL(a.href).origin === location.origin);
const httpLinks = links.map(a => new URL(a.href).protocol === 'http:');
```

I don't think the reuse of `a` is generally seen as bad.

**Comment by annevk** [comment]
Reuse across scopes that are isolated from each other seems okay indeed.
